src/ring_theory/dedekind_domain.lean

```diff
@@ -352,4 +374,216 @@ theorem is_dedekind_domain : is_dedekind_domain A :=
 
 end is_dedekind_domain_inv
 
+variables [algebra A K] [is_fraction_ring A K]
+
+lemma exists_multiset_prod_cons_le_and_prod_not_le [is_dedekind_domain A]
+  (hNF : ¬ is_field A) {I M : ideal A} (hI0 : I ≠ ⊥) (hIM : I ≤ M) [hM : M.is_maximal] :
+  ∃ (Z : multiset (prime_spectrum A)),
+    (M ::ₘ (Z.map prime_spectrum.as_ideal)).prod ≤ I ∧
+    ¬ (multiset.prod (Z.map prime_spectrum.as_ideal) ≤ I) :=
+begin
+  -- Let `Z` be a minimal set of prime ideals such that their product is contained in `J`.
+  obtain ⟨Z₀, hZ₀⟩ := exists_prime_spectrum_prod_le_and_ne_bot_of_domain hNF hI0,
+  obtain ⟨Z, ⟨hZI, hprodZ⟩, h_eraseZ⟩ := multiset.well_founded_lt.has_min
+    (λ Z, (Z.map prime_spectrum.as_ideal).prod ≤ I ∧ (Z.map prime_spectrum.as_ideal).prod ≠ ⊥)
+    ⟨Z₀, hZ₀⟩,
+  have hZM : multiset.prod (Z.map prime_spectrum.as_ideal) ≤ M := le_trans hZI hIM,
+  have hZ0 : Z ≠ 0, { rintro rfl, simpa [hM.ne_top] using hZM },
+  obtain ⟨_, hPZ', hPM⟩ := (hM.is_prime.multiset_prod_le (mt multiset.map_eq_zero.mp hZ0)).mp hZM,
+  -- Then in fact there is a `P ∈ Z` with `P ≤ M`.
+  obtain ⟨P, hPZ, rfl⟩ := multiset.mem_map.mp hPZ',
+  letI := classical.dec_eq (ideal A),
+  have := multiset.map_erase prime_spectrum.as_ideal subtype.coe_injective P Z,
+  obtain ⟨hP0, hZP0⟩ : P.as_ideal ≠ ⊥ ∧ ((Z.erase P).map prime_spectrum.as_ideal).prod ≠ ⊥,
+  { rwa [ne.def, ← multiset.cons_erase hPZ', multiset.prod_cons, ideal.mul_eq_bot,
+         not_or_distrib, ← this] at hprodZ },
+  -- By maximality of `P` and `M`, we have that `P ≤ M` implies `P = M`.
+  have hPM' := (is_dedekind_domain.dimension_le_one _ hP0 P.is_prime).eq_of_le hM.ne_top hPM,
+  tactic.unfreeze_local_instances,
+  subst hPM',
+
+  -- By minimality of `Z`, erasing `P` from `Z` is exactly what we need.
+  refine ⟨Z.erase P, _, _⟩,
+  { convert hZI,
+    rw [this, multiset.cons_erase hPZ'] },
+  { refine λ h, h_eraseZ (Z.erase P) ⟨h, _⟩ (multiset.erase_lt.mpr hPZ),
+    exact hZP0 }
+end
+
+namespace fractional_ideal
+
+lemma exists_not_mem_one_of_ne_bot [is_dedekind_domain A]
+  (hNF : ¬ is_field A) {I : ideal A} (hI0 : I ≠ ⊥) (hI1 : I ≠ ⊤) :
+  ∃ x : K, x ∈ (I⁻¹ : fractional_ideal A⁰ K) ∧ x ∉ (1 : fractional_ideal A⁰ K) :=
+begin
+  -- WlOG, let `I` be maximal.
```

Comment:
Is there a reason why `l` is lowercase but the rest of `wlog` is uppercase?

Reply:
Nope, just a typo.
```suggestion
  -- WLOG, let `I` be maximal.
```